Print extra output from TDSCF / TDDFT calculations

lazaroid opened the pull request:
## Description
This adds further output to TDSCF and TDDFT calculations, including electric TDM vectors and the excitations and deexcitations that contribute to each excited state. Currently, excitations / deexcitations are only printed for C1 symmetry.

A new option TDSCF_COEFF_CUTOFF (default = 0.1) has been added that sets the cutoff for printing excitation / deexcitation coefficients.

## Todos
- [ ] Add support for symmetry
- [ ] Add further output (e.g. other transition dipoles, etc. I've only added what I'm usually after from these calculations)

## Questions
- [ ] How to construct excitations / deexcitations when symmetry is used?
- [ ] Does documentation for new options get built automatically or do I need to add something somewhere?

## Checklist
- [ ] Further comparisons could be added to the tdscf tests, e.g. TDM vectors
- [ ] All tdscf-? tests pass (although the output will be longer than the reference output; other tests still running - slow test machine - but should run as expected) 

## Status
- [x] Ready for review
- [ ] Ready for merge

Zymergen-jvandezande commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -492,6 +493,9 @@ def tdscf_excitations(wfn,
        How verbose should the solver be?
        Default: 1
+    coeff_cutoff : float, optional.
+       Cutoff for printing out excitation and deexcitation coefficients.
```
`deexcitation` should be `de-excitation`. Also in other places in the code.

Zymergen-jvandezande commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -681,4 +685,101 @@ def tda_solver(e, n, g, m):
 
+# Print out TDM vectors just length representation for now: other forms could be added)
+    core.print_out("Electric Transition Dipole Moments (Length) (AU):\nState       X          Y          Z\n")
+    for i, x in enumerate(_results):
+        core.print_out(f" {i+1: 4} {x.edtm_length[0]:< 10.6f} {x.edtm_length[1]:< 10.6f} {x.edtm_length[2]:< 10.6f}\n")
```
Could you unpack x.etdm_length to make this more readable.

e.g.
```
for i, q in enumerate(_results):
    x, y, z = q.edtm_length
    core.print_out(f" {i+1: 4d} {x:< 10.6f} {y:< 10.6f} {z:< 10.6f}\n")
```

lazaroid replied:
Done.

Zymergen-jvandezande reviewed (commented):
This looks awesome, and I'm looking forward to using it!

One additional note, a few of the expressions in f-strings could use formatting to match the PEP8.

robertodr commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -681,4 +685,101 @@ def tda_solver(e, n, g, m):
 
     core.print_out("\n")
 
+# Print out TDM vectors just length representation for now: other forms could be added)
```
The transition moments are not observable: I'd prefer them to be printed only with a high print level. In the output file they're mostly useful only for debugging, I believe.

lazaroid replied:
We use TDM vectors extensively for modelling the alignment of dye molecules within liquid crystal guest-host systems, i.e. how well is a dye transition likely to be aligned with the bulk of the liquid crystal phase.

robertodr reviewed (changes requested):
Thanks for adding this! I believe we can extend it to use symmetry later on. On top of the changes suggested by @Zymergen-jvandezande, I think it's preferable to separate your additions into a separate function, _e.g._ `analyze_tdscf_excitations` which would take an object of `_TDSCFResults` type. The function would also accept a list of transition moments to print.

lazaroid commented:
> Thanks for adding this! I believe we can extend it to use symmetry later on. On top of the changes suggested by @Zymergen-jvandezande, I think it's preferable to separate your additions into a separate function, _e.g._ `analyze_tdscf_excitations` which would take an object of `_TDSCFResults` type. The function would also accept a list of transition moments to print.

lazaroid commented:
Yes, it would certainly make more sense to put this into a separate "analyse" function. I'll try to get that and the other changes done in the next day or so.

hokru commented:
Nice work! Also please don't hesitate to join slack for additional questions/discussions (link in the mainpage header under communication)

lazaroid commented:
I _think_ that's all of the suggestions done (apart from the comment about PEP8 formatting strings: this is the first time I've used f"" strings and so I'm making it up as I go along, to some extent!).

I've moved my new parts to a new function that gets called at the end of tdscf_excitations() (I'm happy to alter it so that it only gets called if verbose > 1, if that would be the preferred option, but I like to see the list of transitions from this sort of calculation.)

I've also added the new array option TDSCF_TDM_PRINT that determines which TDMs are printed (E_TDM_LEN, E_TDM_VEL, and M_TDM begin possible options). This also deals with the reference of @robertodr of not printing TDMs by default.

N.B., the coefficients printed by my code are unscaled but the eigenvectors all seem to be normalised such that X_ssq - Y_ssq = 1 for restricted formalism, and Xalpha_ssq - Yalpha_ssq = 0.5 and Xbeta_ssq - Ybeta_ssq = 0.5.

Coefficients need to be divided by sqrt(2) for comparison with those from The Code That Must Not Be Named, where X_ssq - Y_ssq = 0.5 for restricted formalism (alpha and beta parts may or may not be normalised for unrestricted formalism).

This is why I'm printing the sums of squares in the output, so that users can make a decision based on them, if needed.

robertodr reviewed (commented):
Please also run  the modified file through YAPF, to keep the formatting consistent: https://psicode.netlify.app/psi4manual/master/code_style.html

robertodr commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -425,6 +425,126 @@ def _validate_tdscf(*, wfn, states, triplets, guess) -> None:
+
+    restricted = wfn.same_a_b_orbs()
+
+    if 'E_TDM_LEN' in tdm_print:
+    # Print out ETDM vectors (length representation)
+        core.print_out("\nElectric Transition Dipole Moments (Length) (AU):\nState      X          Y          Z\n")
+        for i, q in enumerate(tdscf_results):
+            x, y, z = q.edtm_length
+            core.print_out(f" {i+1: 4d} {x:< 10.6f} {y:< 10.6f} {z:< 10.6f}\n")
+    if 'E_TDM_VEL' in tdm_print:
+    # Print out ETDM vectors (velocity representation)
+        core.print_out("\nElectric Transition Dipole Moments (Velocity) (AU):\nState      X          Y          Z\n")
+        for i, q in enumerate(tdscf_results):
+            x, y, z = q.edtm_velocity
+            core.print_out(f" {i+1: 4d} {x:< 10.6f} {y:< 10.6f} {z:< 10.6f}\n")
+    if 'M_TDM' in tdm_print:
+    # Print out MTDM vectors
+        core.print_out("\nMagentic Transition Dipole Moments (AU):\nState      X          Y          Z\n")
+        for i, q in enumerate(tdscf_results):
+            x, y, z = q.mdtm
+            core.print_out(f" {i+1: 4d} {x:< 10.6f} {y:< 10.6f} {z:< 10.6f}\n")
+
```
I suggest a slight refactoring here, such that adding other transition moments later doesn't require too much boilerplate:

```python3
_printable = {
"E_TDM_LEN": {"title": "Electric Transition Dipole Moments (Length) (AU)", "what": lambda x: x.edtm_length, },
"E_TDM_VEL": {"title": "Electric Transition Dipole Moments (Velocity) (AU)",  "what": lambda x: x.edtm_velocity, },
"M_TDM": {"title": "Magnetic Transition Dipole Moments (AU)", "what": lambda x: x.mdtm, },
}

for p in tdm_print:
     core.print_out(f"\n{_printable['p']['title']:\nState      X          Y          Z\n")
     for i, q in enumerate(tdscf_results):
          x, y, z = _printable['p']['what'](q)
          core.print_out(f" {i+1: 4d} {x:< 10.6f} {y:< 10.6f} {z:< 10.6f}\n")
```

lazaroid replied:
Oooo...I like that. Done (with a couple of corrections).

I've also run (my new bits of code) through yapf and made the suggested changes (there were a lot suggested for the rest of scf_response.py but that's probably for someone else to sort out).

robertodr commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -492,6 +635,16 @@ def tdscf_excitations(wfn,
        How verbose should the solver be?
        Default: 1
+    coeff_cutoff : float, optional.
+       Cutoff for printing out excitation and deexcitation coefficients.
```
```suggestion
       Cutoff for printing out excitation and de-excitation coefficients.
```

robertodr reviewed (approved):
Thanks again for this!

robertodr commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -425,6 +425,147 @@ def _validate_tdscf(*, wfn, states, triplets, guess) -> None:
+    # Print out requested dipole moment vectors
+    _printable = {
+        "E_TDM_LEN": {
+            "title": "Electric Transition Dipole Moments (Length) (AU)",
```
```suggestion
            "title": "Electric Transition Dipole Moments (Length) (au)",
```

robertodr commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -425,6 +425,147 @@ def _validate_tdscf(*, wfn, states, triplets, guess) -> None:
+            "what": lambda x: x.edtm_length,
+        },
+        "E_TDM_VEL": {
+            "title": "Electric Transition Dipole Moments (Velocity) (AU)",
```
```suggestion
            "title": "Electric Transition Dipole Moments (Velocity) (au)",
```

robertodr commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -425,6 +425,147 @@ def _validate_tdscf(*, wfn, states, triplets, guess) -> None:
+            "what": lambda x: x.edtm_velocity,
+        },
+        "M_TDM": {
+            "title": "Magnetic Transition Dipole Moments (AU)",
```
```suggestion
            "title": "Magnetic Transition Dipole Moments (au)",
```

robertodr commented on psi4/driver/procrouting/response/scf_response.py:
```diff
@@ -425,6 +425,147 @@ def _validate_tdscf(*, wfn, states, triplets, guess) -> None:
+                f"\nExcited State {i+1:4d} ({1 if x.spin_mult== 'singlet' else 3} {x.irrep_ES}):"
+            )
+            core.print_out(
+                f"{x.E_ex_au:> 10.5f} Ha   {E_ex_nm: >.2f} nm f = {x.f_length: >.4f}\n"
```
```suggestion
                f"{x.E_ex_au:> 10.5f} au   {E_ex_nm: >.2f} nm f = {x.f_length: >.4f}\n"
```

lazaroid commented:
Id forgotten all about this! Thanks for checking consistency, etc. :-)

loriab reviewed (approved):
lgtm, thanks!
